### PR TITLE
Cache built deps

### PR DESCRIPTION
**Pull request by arcusfelis**
This PR addresses "we don't need to recompile deps each build".

Proposed changes include:
* Cache binaries of deps based on hash of our configs
* Each job has its own cache
* We would assume, that job uses cache of a parent
* Allow full builds during crons to detect possible problems with missing deps, for example
* Changes in rebar.config or rebar.lock would invalidate cache

### CODE REVIEW

**Comment by mongoose-im** [comment]
[6992.1](https://travis-ci.org/esl/MongooseIM/jobs/569780891) / Erlang 22.0 / small_tests / 602b9f34de402d3cd875ff146f79457d2b5f3018
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2411/6992/small_tests.22.0) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2411/6992/small_tests.22.0/small/ct_run.mongooseim@localhost.2019-08-09_12.42.10/lib.mongooseim.logs/run.2019-08-09_12.42.10/suite.log.html)

**Comment by mongoose-im** [comment]
[6996.1](https://travis-ci.org/esl/MongooseIM/jobs/570183853) / Erlang 22.0 / small_tests / d123eecd6824f8e516fefb1473a5ae2a9d13909b
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2411/6996/small_tests.22.0) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2411/6996/small_tests.22.0/small/ct_run.mongooseim@localhost.2019-08-10_14.02.51/lib.mongooseim.logs/run.2019-08-10_14.02.51/suite.log.html)

File: tools/cache-build.sh [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+# Travis packs all cache directories into one archive for caching.
+# So, we put files into tar, not tar.gz to improve compression
+# (because the files are shared between mongooseim and big_tests)
+
+set -e
+CACHE_DIR=~/.cache/mim_builds
+
+## Use array assignment to get just hash without filename
+## https://stackoverflow.com/a/5773761
+SHA_ARRAY=($(cat "rebar.config" "rebar.lock" | shasum))
```
Comment (ludwikbukowski) [inline tools/cache-build.sh]: what about `big_tets/rebar.config` ?

Comment (arcusfelis) [reply tools/cache-build.sh]: There is tools/cache-tests.sh for caching big_tests/_build.

Comment (ludwikbukowski) [reply tools/cache-build.sh]: sorry, you're right, I didnt see it at this point of review :D

**Pull request merged**